File: dotnet/sql_injection/ProductQueries.cs
using System.Collections.Generic;
using System.Data.SqlClient;

public class ProductQueries
{
    private readonly string _connectionString = "Server=prod-db;Database=shopist;User Id=app;Password=apppass;";

    // VULN 1: String concatenation SQL injection - product search
    public IEnumerable<Product> SearchProducts(string searchTerm)
    {
        using var conn = new SqlConnection(_connectionString);
        conn.Open();
        string query = "SELECT id, name, price, stock FROM products WHERE name LIKE '%" + searchTerm + "%' OR description LIKE '%" + searchTerm + "%'";
        using var cmd = new SqlCommand(query, conn);
        using var reader = cmd.ExecuteReader();
        return ReadProducts(reader);
    }

    // VULN 2: String interpolation SQL injection - price range filter
    public IEnumerable<Product> GetProductsByPriceRange(string minPrice, string maxPrice)
    {
        using var conn = new SqlConnection(_connectionString);
        conn.Open();
        string query = $"SELECT * FROM products WHERE price BETWEEN {minPrice} AND {maxPrice} ORDER BY price ASC";
        using var cmd = new SqlCommand(query, conn);
        using var reader = cmd.ExecuteReader();
        return ReadProducts(reader);
    }

    // VULN 3: String concatenation with ORDER BY injection - category filter
    public IEnumerable<Product> GetProductsByCategory(string category, string sortField)
    {
        using var conn = new SqlConnection(_connectionString);
        conn.Open();
        string query = "SELECT * FROM products WHERE category = '" + category + "' ORDER BY " + sortField;
        using var cmd = new SqlCommand(query, conn);
        using var reader = cmd.ExecuteReader();
        return ReadProducts(reader);
    }

    private IEnumerable<Product> ReadProducts(SqlDataReader reader) => new List<Product>();
}

public class Product { public int ID, Stock; public string Name, Category; public decimal Price; }

File: ruby/sql_injection/product_queries.rb
require 'active_record'

# VULN 1: ActiveRecord string interpolation in where - product search
def search_products(search_term)
  Product.where("name LIKE '%#{search_term}%' OR description LIKE '%#{search_term}%'")
end

# VULN 2: String concatenation in where - price range filter
def get_products_by_price_range(min_price, max_price)
  Product.where("price BETWEEN " + min_price.to_s + " AND " + max_price.to_s)
end

# VULN 3: find_by_sql with string interpolation and ORDER BY injection
def get_products_by_category(category, sort_field)
  Product.find_by_sql(
    "SELECT * FROM products WHERE category = '#{category}' ORDER BY #{sort_field}"
  )
end
